### PR TITLE
This adds a `@nest` member to term definitions used for _transparent properties_

### LINKED ISSUE #246: Ignoring semantically meaningless nesting
Thanks for the great work with JSON-LD! However, when trying to use JSON-LD for to present data in the company I'm working in, I noticed the following missing feature:

FEATURE PROPOSAL: ABILITY TO DEFINE ANY KEY AS AN INDEX KEY

In addition to JSON-LD's existing index container structure, I propose that any key under a JSON-LD node could be defined as a index key.

This would help clustering data under a node into coder friendly logical groups without messing up the Linked Data interpretation with e.g. blank nodes. I encountered the need for this feature at our company where our problem is that the amount of attributes a single JSON-LD node can have can potentially be quite many, say, tens or hundreds of attributes.

As far as I know, this can not be currently done with JSON-LD without 1) ending up with blank nodes or  2) the need to create a deeper JSON structure by using a separate index term (using "@container":"@index") which then contains the data underneath.

In addition, if a single key could be defined as a index term, this would make it more flexible to attach the JSON-LD Linked Data interpretation to even a wider amount of existing JSON data, without having to change the structure of such data (and without ending up with e.g. lots of blank nodes).

DEFINING AN INDIVIDUAL INDEX KEY IN @CONTEXT

The "@context" definition could be done e.g. using the existing reserved keyword "@index" in the following way:

```
"indexkey":"@index"
```

which should be interpreted in the following way: 1) the "indexkey" is an index key and should be skipped when traversing the JSON tree while doing the JSON-LD to RDF interpretation, 2) any data directly under the "indexkey" should be interpreted as data directly attached to the node of the indexkey (same RDF subject).

EXAMPLE

To give a full example, in the following a single key "labels" is defined as an index index key to help grouping the data into coder friendly logical groups without messing up the Linked Data interpretation):

``` js
{
  "@context": {
     "labels":"@index",
     "main_label":"http://example.org/my-schema#main_label",
     "other_label":"http://example.org/my-schema#other_label",
     "homepage":{ "@id":"http://example.org/my-schema#homepage", "@type":"@id"}
  },
  "@id":"http://example.org/myresource",
  "homepage": "http://example.org",
  "labels": {
     "main_label": "This is the main label for my resource",
     "other_label": "This is the other label"
  }
}
```

This example JSON-LD should generate the following RDF triplets:

```
<http://example.org/myresource> <http://example.org/my-schema#homepage> <http://example.org>. 
<http://example.org/myresource> <http://example.org/my-schema#main_label> "This is the main label for my resource".
<http://example.org/myresource> <http://example.org/my-schema#other_label> "This is the other label".
```

### PR DESCRIPTION
**Pull request by gkellogg**
This introduces the `@nest` keyword, used for folding contained properties to the enclosing node object, and allows for `@nest` in a term definition, to introduce such nesting when compacting.

Fixes #246. This also is an alternative to PR #451 and is (mostly) independent of #449).

In contrast to #451, having a parallel `@nest` term definition property allows this to be used in conjunction with `@container` (but not for `@reverse` compaction).

There are several new expand, compact and error tests for this feature.

Formatted HTML versions are available for [JSON-LD](http://gkellogg.github.io/json-ld.org/spec/latest/json-ld/), and [JSON-LD-API](http://gkellogg.github.io/json-ld.org/spec/latest/json-ld-api/) on my personal fork.

### CODE REVIEW

File: spec/latest/json-ld-api/index.html [hunk 1]
```diff
@@ -4058,6 +4112,7 @@
             "invalid language mapping",
             "invalid language-tagged string",
             "invalid language-tagged value",
+            "invalid @nest value",
```
Comment (davidlehn) [inline spec/latest/json-ld-api/index.html]: move up sorted with other @ values

File: spec/latest/json-ld-api/index.html [hunk 2]
```diff
@@ -4134,6 +4189,8 @@
           associated language tag was detected.</dd>
         <dt><dfn>invalid local context</dfn></dt>
         <dd>In invalid <a>local context</a> was detected.</dd>
+        <dt class="changed"><dfn>invalid @nest value</dfn></dt>
```
Comment (davidlehn) [inline spec/latest/json-ld-api/index.html]: move up sorted with other @ values

File: spec/latest/json-ld-api/index.html (hunk 1 above)
Comment (gkellogg) [inline spec/latest/json-ld-api/index.html]: Good call, the `@` was added later. Should be good now.

**Comment by gkellogg** [comment]
@dbooth-boston Flattening is an entirely different concept, which takes nested node definitions and flattens them so that nodes only reference other nodes, rather than include them.

@dlongley suggested `@preserve`, but that has a different use in framing.

However, the use case you may be looking for could be handled using scoped contexts, there interpretation of "location" is different for "file" vs "vehicle". This was the subject of a recently merged change. But, that does imply that both "file" and "vehicle" refer to different entities, rather than are each properties of the containing node, which is more consistent with what I think you want.

As for a `@nest` alternative, I'm open to suggestions.

**Comment by dbooth-boston** [comment]
I was thinking of _.flatten from http://underscorejs.org/#flatten -- not the JSON-LD concept of flattening. I should not have used that word.  Sorry for the confusion!

Yes, the use case I was mentioning would be the ability to distinguish between properties (of the same name) by their object ancestors, as illustrated above.  This came up a couple of years ago when we were originally trying to see if FHIR/JSON could be adapted to be JSON-LD.  We concluded that it could not without either changing the existing property names or inserting additional @contexts into the instance data -- neither of which would be palatable to the vanilla FHIR/JSON folks.

**Pull request merged**